packages/docusaurus-theme-common/src/contexts/tabGroupChoice.tsx

```diff
@@ -13,61 +13,123 @@ import React, {
   useContext,
   type ReactNode,
 } from 'react';
+import {useHistory, useLocation} from '@docusaurus/router';
 import {createStorageSlot, listStorageKeys} from '../utils/storageUtils';
 import {ReactContextError} from '../utils/reactUtils';
 
 const TAB_CHOICE_PREFIX = 'docusaurus.tab.';
 
 type ContextValue = {
-  /** A boolean that tells if choices have already been restored from storage */
-  readonly ready: boolean;
-  /** A map from `groupId` to the `value` of the saved choice. */
-  readonly tabGroupChoices: {readonly [groupId: string]: string};
+  /** A map from `groupId` to the `value` of the saved choice in storage. */
+  readonly tabGroupChoicesInStorage: {readonly [groupId: string]: string};
+  /** A map from `searchKey` to the `value` of the choice in query parameter. */
+  readonly tabGroupChoicesInQueryParams: {readonly [searchKey: string]: string};
```

Comment:
- Remove `ready` property.
- Manage both local storage values & query param values in this context.